### ISSUE TITLE
Error when overriding profile.tpl.php 

### ISSUE THREAD

**Issue by robertgarrigos** [backdrop-contrib/profile#35]
Overwriting profile.tpl.php triggers an error:

```
Warning: Undefined array key "#entity" a template_preprocess_entity_plus() (línia 147 de /Users/robert/Sites/backdrop/modules/entity_plus/theme/entity_plus.theme.inc).
Error: Call to a member function label() on null a entity_label() (línia 774 de /Users/robert/Sites/backdrop/core/modules/entity/entity.module).
```
I found the same kind of error with paragraphs (https://github.com/backdrop-contrib/paragraphs/issues/91) which has been fixed.

**Comment by argiepiano** [comment]
What do you mean by "overwriting"? Overriding? 

Can you add more specific steps to reproduce? I believe you have Entity Plus enabled (which is not required for this module).

I do see some issues in `profile.info.inc` which is only used when E+ is enabled. For starters, it has the wrong callbacks: `entity_property_verbatim_set` and `entity_property_getter_method`, which don't exist. In Backdrop's Entity Plus, those should be `entity_plus_property_verbatim_set` and `entity_plus_property_getter_method`. So perhaps this is in part producing this problem (though I doubt it). 

I see another issue, which may also be causing this problem: `profile_entity_info()` defines a `'label callback' => 'entity_class_label'`, which doesn't exist in Backdrop. It should probably be `'entity_label'`.

If you want to try patching your site with those and seeing if that helps, that'd be helpful to pinpoint the problem.

**Comment by robertgarrigos** [comment]
Yes, overriding I meant. I need Entity Plus for some other module. Thanks, @argiepiano, those changes didn't fix the problem. Steps are easy, just create a profile and try to override profile.tpl.php file within your theme. Accessing the profile page triggers the error.

**Comment by argiepiano** [comment]
Did you try clearing caches after the changes? If so, OK, that means that more debugging is needed.

**Comment by robertgarrigos** [comment]
yep, cache was cleared.

**Comment by robertgarrigos** [comment]
Commenting the function profile_entity_info_alter(&$entity_info) in profile.module fixes the error. I have been doing actions on profiles with no problem. Is that hook_entity_info_alter() function used at all?

**Comment by argiepiano** [comment]
~~Yes, it defines metadata for all of profile properties. You won't be able to use entity_metadata_wrapper with profile entities now. But this also points at the fact that the error is there. Most likely an inexistent callback, like I said above. You probably missed some of those.~~

~~To find the culprit, you should look at ALL callbacks specified there (as in `setter callback` , `getter callback` and others) and check that the callback function actually exists.~~

**Comment by argiepiano** [comment]
Actually, what I just sald above is wrong. `profile_entity_info_alter()` does not provide metadata. It actually tells your site to use the controller provided by Entity Plus.

So, commenting that out will avoid using E+ entity controller for Profiles in your site. This is an acceptable solution in your case. But the bug needs to be found, since it still exists.

**Comment by robertgarrigos** [comment]
I'm lost with this bug.

**Comment by argiepiano** [comment]
If I had the time I'd try to reproduce ... not right now unfortunately.

**Comment by bugfolder** [comment]
I can reproduce this. Created a profile, then added profile.tpl.php to Basis and cleared caches. Immediate error screen.

 Now looking into it...

**Comment by argiepiano** [comment]
This problem happens because `template_preprocess_entity_plus()` expects the profile entity to be stored under `#entity`. In D7, this was done under `DefaultEntityController::view()`, the same as in B's `EntityPlusController::view()`. The problem here is that `EntityPlusController::view()` is never invoked - any calls to `entity_view()` now call Backdrop's core controller's `view()` method.

The solution is simple. We just need to pass the entity under `#entity` within `Profile:view()`. @bugfolder basically:

```php
    $build += array(
      '#theme' => 'profile',
      '#profile' => $this,
      '#entity' => $this,
      '#view_mode' => $view_mode,
      '#langcode' => $langcode,
    );
``` 

notice the line `#entity`. This is in Profile::view().

I can provide a PR if you want

**Comment by bugfolder** [comment]
If we comment out the change of entity controller (as described above) and then view a user profile, we end up calling `template_preprocess_profile($variables)`, and all is well.

However, if we leave the change of entity controller to use E+'s controller, then we don't call that preprocess function; instead, we call `template_preprocess_entity_plus($variables)`, and get the crash. This is fundamentally because on line 147 of entity_plus.theme.inc, we have
```
  $entity = $variables['elements']['#entity'];
```
and there is no `#entity` element in `$variables`. The entity object actually does reside in `$variables`, but under the key `#profile`.

This seems to be set in `Profile->view()`, where we set up the array with this code at line 148 of profile.entity.inc:
```
    $build += array(
      '#theme' => 'profile',
      '#profile' => $this,
      '#view_mode' => $view_mode,
      '#langcode' => $langcode,
    );
```

Since `$this` is the entity object in question, I tried adding another copy of it to the array with the `#entity` key:
```
    $build += array(
      '#theme' => 'profile',
      '#profile' => $this,
      '#entity' => $this,
      '#view_mode' => $view_mode,
      '#langcode' => $langcode,
    );
```
This completely fixed the crash, and the user profile rendered correctly.

@argiepiano, based on your considerable entity guru-ship, does this sound like the appropriate change to make Profile module play nicely with E+?

**Comment by bugfolder** [comment]
Hah. Jinx.

**Comment by bugfolder** [comment]
Next time, I won't spend so much time writing up what I found!
